internal/npm_install/generate_build_file.ts

```diff
@@ -0,0 +1,1140 @@
+        executables.set(pkg._dir, cleanupBinPath(pkg.bin));
+      } else if (typeof pkg.bin === 'object') {
+        for (let key in pkg.bin) {
+          if (isValidBinPathStringValue(pkg.bin[key])) {
+            executables.set(key, cleanupBinPath(pkg.bin[key]));
+          }
+        }
+      }
+    }
+  }
+
+  return executables;
+}
+
+// Handle additionalAttributes of format:
+// ```
+// "bazelBin": {
+//   "ngc-wrapped": {
+//     "additionalAttributes": {
+//       "configuration_env_vars": "[\"compile\"]"
+//   }
+// },
+// ```
+function additionalAttributes(pkg: Dep, name: string) {
+  let additionalAttributes = '';
+  if (pkg.bazelBin && pkg.bazelBin[name] && pkg.bazelBin[name].additionalAttributes) {
+    const attrs = pkg.bazelBin[name].additionalAttributes;
+    for (const attrName of Object.keys(attrs)) {
+      const attrValue = attrs[attrName];
+      additionalAttributes += `\n    ${attrName} = ${attrValue},`;
+    }
+  }
+  return additionalAttributes;
+}
+
+/**
+ * Given a pkg, return the skylark nodejs_binary targets for the package.
+ */
+function printPackageBin(pkg: Dep) {
+  let result = '';
+  const executables = _findExecutables(pkg);
+  if (executables.size) {
+    result = `load("@build_bazel_rules_nodejs//:index.bzl", "nodejs_binary")
+
+`;
+    const data = [`//${pkg._dir}:${pkg._name}`];
+    if (pkg._dynamicDependencies) {
+      data.push(...pkg._dynamicDependencies);
+    }
+
+    for (const [name, path] of executables.entries()) {
+      result += `# Wire up the \`bin\` entry \`${name}\`
+nodejs_binary(
+    name = "${name}",
+    entry_point = "//:node_modules/${pkg._dir}/${path}",
+    install_source_map_support = False,
+    data = [${data.map(p => `"${p}"`).join(', ')}],${additionalAttributes(pkg, name)}
+)
+
+`;
+    }
+  }
+
+  return result;
+}
+
+function printIndexBzl(pkg: Dep) {
+  let result = '';
+  const executables = _findExecutables(pkg);
+  if (executables.size) {
+    result = `load("@build_bazel_rules_nodejs//:index.bzl", "nodejs_binary", "npm_package_bin")
+
+`;
+    const data = [`@${WORKSPACE}//${pkg._dir}:${pkg._name}`];
+    if (pkg._dynamicDependencies) {
+      data.push(...pkg._dynamicDependencies);
+    }
+
+    for (const [name, path] of executables.entries()) {
+      result = `${result}
+
+# Generated helper macro to call ${name}
+def ${name.replace(/-/g, '_')}(**kwargs):
+    output_dir = kwargs.pop("output_dir", False)
+    if "outs" in kwargs or output_dir:
+        npm_package_bin(tool = "@${WORKSPACE}//${pkg._dir}/bin:${
+          name}", output_dir = output_dir, **kwargs)
+    else:
+        nodejs_binary(
+            entry_point = "@${WORKSPACE}//:node_modules/${pkg._dir}/${path}",
+            install_source_map_support = False,
+            data = [${data.map(p => `"${p}"`).join(', ')}] + kwargs.pop("data", []),${
+          additionalAttributes(pkg, name)}
+            **kwargs
+        )
+  `;
+    }
+  }
+  return result;
+}
+
+type Dep = {
```

Comment:
should be an interface

Reply:
Interfaces can be extended. Because this type is not exported there isn't a benefit that i know of.

i'm sure there is a style guide but somewhere about this but i'd like to only make style changes if they're backed up by documentation.

also this type is just popped in the middle of the file. it might make sense to move it to another file that just exports interfaces and clean things up.

Reply:
Fair point, Im pretty sure there's some docs for it somewhere but I cant find it from a quick search I just did.
Either way, for me it's just tidier, but you're right there's no functional difference in this situation aside from possible extension later on, but you could just refactor it then.

Reply:
there's a google typescript style guide which we ought to nominally follow